File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/F-Астроград/solution.cpp
#include <iostream>
#include <deque>
#include <unordered_map>

using namespace std;


void solution() {
    /*
    Time Complexity O(1) for each operation
    Memory Complexity O(N)
    */
    int n;
    scanf("%d", &n);
    deque<int> queue;
    unordered_map<int, pair<int, int>> hashmap;
    int leaved = 0;
    for (int i = 0; i < n; ++i) {
        int event;
        scanf("%d", &event);
        if (event == 1) {
            int manId;
            scanf("%d", &manId);
            queue.push_back(manId);
            hashmap[manId] = {queue.size(), leaved};
        } else if (event == 2) {
            hashmap.erase(queue.front());
            queue.pop_front();
            ++leaved;
        } else if (event == 3) {
            hashmap.erase(queue.back());
            queue.pop_back();
        } else if (event == 4) {
            int q;
            scanf("%d", &q);
            printf("%d\n", hashmap[q].first - (leaved - hashmap[q].second) - 1);
        } else {
            printf("%d\n", queue.front());
        }
    }
}

int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/F-Астроград/solution.py
import sys
from collections import deque


def solution() -> None:
    """
    Time Complexity O(1) for each operation
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    queue: deque[int] = deque()
    leaved: int = 0
    hashmap: dict[int, tuple[int, int]] = dict()
    for _ in range(n):
        event, *args = sys.stdin.readline().rstrip().split()
        if event == '1':
            queue.append(args[0])
            hashmap[args[0]] = (len(queue), leaved)
        elif event == '2':
            del hashmap[queue.popleft()]
            leaved += 1
        elif event == '3':
            del hashmap[queue.pop()]
        elif event == '4':
            sys.stdout.write(f'{hashmap[args[0]][0] - (leaved - hashmap[args[0]][1]) - 1}\n')
        else:
            sys.stdout.write(f'{queue[0]}\n')


if __name__ == '__main__':
    solution()